python_src/src/lamp_py/performance_manager/l0_rt_trip_updates.py

```diff
@@ -156,26 +113,22 @@ def get_and_unwrap_tu_dataframe(
             .astype("Int64")
         )
 
-        # expand and filter stop_time_update column using numpy vectorize
-        # numpy vectorize offers significantly better performance over pandas apply
-        # this will return a ndarray with values being list of dicts
-        vector_explode = numpy.vectorize(explode_stop_time_update)
-        batch_events = pandas.Series(
-            vector_explode(
-                batch_events.stop_time_update,
-                batch_events.timestamp,
-                batch_events.direction_id,
-                batch_events.route_id,
-                batch_events.service_date,
-                batch_events.start_time,
-                batch_events.vehicle_id,
-                batch_events.trip_id,
+        batch_events["tu_stop_timestamp"] = pandas.to_numeric(
+            batch_events["tu_stop_timestamp"]
+        ).astype("Int64")
+
+        # filter out stop event predictions that are too far into the future
+        # and are unlikely to be used as a final stop event prediction
+        # (2 minutes) or predictions that go into the past (negative values)
+        batch_events = batch_events[
+            (batch_events["tu_stop_timestamp"] - batch_events["timestamp"] >= 0)
+            & (
+                batch_events["tu_stop_timestamp"] - batch_events["timestamp"]
+                < 120
```

Comment:
explode no longer needed, this filter will limit batch events to predictions that occurred within 2 minutes of actual arrival time.